File: src/lib/solana/pda.ts
/**
 * PDA seed construction, TypeScript side.
 *
 * Twin of `programs/buildshare/src/pda.rs`. Both files must produce
 * byte-identical seeds or the client and the program disagree about which
 * account an operation refers to.
 *
 * Frozen encoding rules (DESIGN FREEZE v1.2):
 *   - seed prefixes are ASCII, exactly as written in constants.rs
 *   - u64 seeds are 8 bytes LITTLE-endian
 *   - the attempt seed is ONE raw byte (u8), never a decimal string
 *   - seed order is fixed and never reordered
 *
 * Dependency-free on purpose: no @solana/web3.js import, so these helpers and
 * their tests run in the demo path and in plain Node.
 */

export const SEED_PROJECT = 'project';
export const SEED_TASK = 'task';
export const SEED_CONTRIBUTION = 'contribution';
export const SEED_MEMBER = 'member';

export const U64_MAX = 18446744073709551615n;

/** ASCII seed prefix as bytes. */
export function asciiSeed(value: string): Uint8Array {
  return new TextEncoder().encode(value);
}

/**
 * 8-byte little-endian encoding of a u64, matching Rust `u64::to_le_bytes()`.
 * BigInt + DataView on purpose: bit shifts in JavaScript are 32-bit and would
 * silently corrupt ids above 2^31.
 */
export function u64le(value: number | bigint): Uint8Array {
  if (typeof value === 'number' && !Number.isInteger(value)) {
    throw new RangeError('u64 seed must be an integer: ' + String(value));
  }
  const big = typeof value === 'bigint' ? value : BigInt(value);
  if (big < 0n || big > U64_MAX) {
    throw new RangeError('u64 seed out of range: ' + String(value));
  }
  const bytes = new Uint8Array(8);
  new DataView(bytes.buffer).setBigUint64(0, big, true);
  return bytes;
}

/** Single raw byte, matching a Rust `&[attempt]` seed where attempt: u8. */
export function u8byte(value: number): Uint8Array {
  if (!Number.isInteger(value) || value < 0 || value > 255) {
    throw new RangeError('u8 seed out of range: ' + String(value));
  }
  return new Uint8Array([value]);
}

export function projectSeeds(
  founderBytes: Uint8Array,
  onchainProjectId: number | bigint,
): Uint8Array[] {
  return [asciiSeed(SEED_PROJECT), founderBytes, u64le(onchainProjectId)];
}

export function taskSeeds(
  projectPdaBytes: Uint8Array,
  onchainTaskId: number | bigint,
): Uint8Array[] {
  return [asciiSeed(SEED_TASK), projectPdaBytes, u64le(onchainTaskId)];
}

export function contributionSeeds(
  taskPdaBytes: Uint8Array,
  contributorBytes: Uint8Array,
  attempt: number,
): Uint8Array[] {
  return [asciiSeed(SEED_CONTRIBUTION), taskPdaBytes, contributorBytes, u8byte(attempt)];
}

export function memberSeeds(
  projectPdaBytes: Uint8Array,
  walletBytes: Uint8Array,
): Uint8Array[] {
  return [asciiSeed(SEED_MEMBER), projectPdaBytes, walletBytes];
}

/** Debug/parity helper: lowercase hex of a seed list, joined by '|'. */
export function seedsToHex(seeds: Uint8Array[]): string {
  return seeds
    .map((seed) =>
      Array.from(seed)
        .map((b) => b.toString(16).padStart(2, '0'))
        .join(''),
    )
    .join('|');
}

File: programs/buildshare/src/pda.rs
use crate::constants::*;
use anchor_lang::prelude::*;

/// Address derivation, kept in one place so that Rust and TypeScript can be
/// diffed against a single definition. The TypeScript twin lives in
/// `src/lib/solana/pda.ts` and is covered by byte-vector parity tests.
///
/// Encoding rules, frozen:
///   * `u64` seeds are 8-byte LITTLE-endian (`to_le_bytes`)
///   * `u8` seeds are a single raw byte, never a decimal string
///   * seed order is fixed and never reordered
pub fn project_pda(program_id: &Pubkey, founder: &Pubkey, project_id: u64) -> (Pubkey, u8) {
    Pubkey::find_program_address(
        &[SEED_PROJECT, founder.as_ref(), &project_id.to_le_bytes()],
        program_id,
    )
}

pub fn task_pda(program_id: &Pubkey, project: &Pubkey, task_id: u64) -> (Pubkey, u8) {
    Pubkey::find_program_address(
        &[SEED_TASK, project.as_ref(), &task_id.to_le_bytes()],
        program_id,
    )
}

pub fn contribution_pda(
    program_id: &Pubkey,
    task: &Pubkey,
    contributor: &Pubkey,
    attempt: u8,
) -> (Pubkey, u8) {
    Pubkey::find_program_address(
        &[SEED_CONTRIBUTION, task.as_ref(), contributor.as_ref(), &[attempt]],
        program_id,
    )
}

pub fn member_pda(program_id: &Pubkey, project: &Pubkey, wallet: &Pubkey) -> (Pubkey, u8) {
    Pubkey::find_program_address(
        &[SEED_MEMBER, project.as_ref(), wallet.as_ref()],
        program_id,
    )
}

#[cfg(test)]
mod tests {
    use super::*;

    fn program() -> Pubkey {
        Pubkey::new_from_array([7u8; 32])
    }

    fn founder() -> Pubkey {
        Pubkey::new_from_array([1u8; 32])
    }

    fn contributor() -> Pubkey {
        Pubkey::new_from_array([2u8; 32])
    }

    #[test]
    fn project_pda_is_deterministic() {
        let a = project_pda(&program(), &founder(), 1);
        let b = project_pda(&program(), &founder(), 1);
        assert_eq!(a, b);
    }

    #[test]
    fn project_id_changes_the_address() {
        let a = project_pda(&program(), &founder(), 1).0;
        let b = project_pda(&program(), &founder(), 2).0;
        assert_ne!(a, b);
    }

    #[test]
    fn founder_changes_the_address() {
        let a = project_pda(&program(), &founder(), 1).0;
        let b = project_pda(&program(), &contributor(), 1).0;
        assert_ne!(a, b);
    }

    #[test]
    fn u64_seed_is_little_endian() {
        assert_eq!(1u64.to_le_bytes(), [1, 0, 0, 0, 0, 0, 0, 0]);
        assert_eq!(
            258u64.to_le_bytes(),
            [2, 1, 0, 0, 0, 0, 0, 0],
            "258 = 0x0102 must serialise low byte first"
        );
    }

    #[test]
    fn attempt_seed_is_one_raw_byte() {
        let project = project_pda(&program(), &founder(), 1).0;
        let task = task_pda(&program(), &project, 1).0;
        let a1 = contribution_pda(&program(), &task, &contributor(), 1).0;
        let a2 = contribution_pda(&program(), &task, &contributor(), 2).0;
        assert_ne!(a1, a2, "each attempt must get its own contribution account");
    }

    #[test]
    fn member_address_depends_on_project_and_wallet() {
        let p1 = project_pda(&program(), &founder(), 1).0;
        let p2 = project_pda(&program(), &founder(), 2).0;
        assert_ne!(
            member_pda(&program(), &p1, &contributor()).0,
            member_pda(&program(), &p2, &contributor()).0
        );
    }
}
